native/explorer/src/datatypes.rs

```diff
@@ -321,14 +321,63 @@ fn convert_to_arrow_time_unit(time_unit: TimeUnit) -> ArrowTimeUnit {
     }
 }
 
-// Limit the number of digits in the microsecond part of a timestamp to 6.
-// This is necessary because the microsecond part of Elixir is only 6 digits.
-#[inline]
-pub fn microseconds_six_digits(microseconds: u32) -> u32 {
-    if microseconds > 999_999 {
-        999_999
-    } else {
-        microseconds
+/// Trait for the fractional part of seconds for Elixir's time-related structs.
+pub trait ExMicrosecondTuple {
+    /// Get the sub-second part.
+    fn subsec_micros(&self) -> u32;
+
+    /// Round the sub-second part to a max of 6 digits.
+    fn subsec_micros_limited(&self) -> u32 {
+        // In the event of a leap second, `subsec_micros()` may exceed 999,999.
+        self.subsec_micros().min(999_999_u32)
+    }
+
+    /// Default representation.
+    fn microsecond_tuple(&self) -> (u32, u32) {
+        (self.subsec_micros_limited(), 6)
+    }
+
+    /// Representation rounded for specific time-units.
+    fn microsecond_tuple_tu(&self, time_unit: TimeUnit) -> (u32, u32) {
+        let us = self.subsec_micros_limited();
+
+        match time_unit {
+            // If Polars implements `TimeUnit::Seconds`, we'd return `(0, 0)`.
+            TimeUnit::Microseconds | TimeUnit::Nanoseconds => (us, 6),
+            TimeUnit::Milliseconds => (((us / 1_000) * 1_000), 3),
+        }
+    }
+}
+
+impl ExMicrosecondTuple for NaiveTime {
+    fn subsec_micros(&self) -> u32 {
+        self.nanosecond() / 1_000
+    }
+    // fn microsecond_tuple(&self) -> (u32, u32) {
+    //     let us = self.subsec_micros_limited();
+    //     if us == 0 {
+    //         (0, 0)
+    //     } else {
+    //         (self.subsec_micros_limited(), 6)
+    //     }
+    // }
```

Comment:
There's currently a test that fails with:

```elixir
       iex> s = Explorer.Series.from_list([~T[01:55:00], ~T[15:35:00], ~T[23:00:00]])
       iex> Explorer.Series.quantile(s, 0.5)
       ~T[15:35:00]
     code:  Explorer.Series.quantile(s, 0.5) === ~T[15:35:00]
     left:  ~T[15:35:00.000000]
     right: ~T[15:35:00]
```

The commented code would fix this test. However, I think we should change the test instead.

The Rust-side can't tell the difference between a `chrono::NaiveTime` that was originally encoded from an Elixir `%Time{}` with `:second`-precision and one with higher precision that happens to have 0 microseconds. I think the best we can do is return what we know for sure.

Thoughts?

Reply:
We can change the test. All of the times given in the test have precision of 0, so ideally we would return precision 0, but if we have to give it higher precision, any of the precisions are fine.

Reply:
> ideally we would return precision 0

Right. Unfortunately there's no way to annotate this now because of the lack of `:second` precision in Polars: https://github.com/elixir-explorer/explorer/issues/885. Also we only support `:time` and not `{:time, :millisecond | :microsecond | :nanosecond}` (though we could if we wanted).